### PR TITLE
Fix Azure Monitor support for multiple aggregation types

### PR DESCRIPTION
**Pull request by zmoog**
## Proposed commit message

Fixes two bugs related to supporting multiple aggregation types.

1. Use aggregation type in `MetricRegistry` in the cache key.
2. Create one `KeyValuePoint` for each aggregation type

### (1) Use aggregation type in `MetricRegistry` in the cache key

The `MetricRegistry.NeedsUpdate` checks if a metric definition needs an update. If there are multiple metrics config with no dimensions:

```yaml
  resources:
  - resource_query: "resourceType eq 'Microsoft.EventHub/Namespaces'"
    metrics:
    - aggregations:
        - Maximum
      ignore_unsupported: true
      name:
        - ActiveConnections
      namespace: Microsoft.EventHub/Namespaces
    - aggregations:
        - Minimum
      ignore_unsupported: true
      name:
        - ActiveConnections
      namespace: Microsoft.EventHub/Namespaces
    - aggregations:
        - Average
      ignore_unsupported: true
      name:
        - ActiveConnections
      namespace: Microsoft.EventHub/Namespaces
```

All three metric configs end up having the same key, ending up skipping the metric collection for the second and third configs.

By adding the aggregation type in the cache key, we can correctly keep track of metric updates.

### (2) Create one `KeyValuePoint` for each aggregation type

Add support for multiple aggregation types in the `mapToKeyValuePoints()` function to restore support for configs like the following:

```yaml
  resources:
  - resource_query: "resourceType eq 'Microsoft.EventHub/Namespaces'"
    metrics:
    - namespace: Microsoft.EventHub/Namespaces
      name:
        - ActiveConnections
      aggregations:
        - Average
        - Maximum
        - Minimum
      ignore_unsupported: true
      timegrain: PT1M
```

I missed this feature in a previous change. Unfortunately, no existing modules or integrations use multiple aggregation types, and no tests exist.

We are adding more tests targeting to cover all existing use cases soon.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] ~~I have made corresponding changes to the documentation~~
- [ ] ~~I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [ ] Cover all multiple aggregation type use cases with tests

## Related issues

- Closes https://github.com/elastic/beats/issues/39192

## Use cases

The PR must support multiple aggregation types in two different variants:

- multiple aggregation types in one metric group
- one aggregation type in multiple groups with the same metric

Here are two variants of the metric `ActiveConnections` with three aggregation types: `Maximum`, `Minimum`, and `Avergate`:

```yaml
- module: azure
  metricsets:
    - monitor
  enabled: true
  period: 60s
  client_id: '<redacted>'
  client_secret: '<redacted>'
  tenant_id: '<redacted>'
  subscription_id: '<redacted>'
  refresh_list_interval: 600s
  resources:
  - resource_query: "resourceType eq 'Microsoft.EventHub/Namespaces'"
    metrics:
    - aggregations:
        - Maximum
      ignore_unsupported: true
      name:
        - ActiveConnections
      namespace: Microsoft.EventHub/Namespaces
    - aggregations:
        - Minimum
      ignore_unsupported: true
      name:
        - ActiveConnections
      namespace: Microsoft.EventHub/Namespaces
    - aggregations:
        - Average
      ignore_unsupported: true
      name:
        - ActiveConnections
      namespace: Microsoft.EventHub/Namespaces
```

This variant should be preferred since in needs less API calls to fetch the same data:

```yaml
- module: azure
  metricsets:
    - monitor
  enabled: true
  period: 60s
  client_id: '<redacted>'
  client_secret: '<redacted>'
  tenant_id: '<redacted>'
  subscription_id: '<redacted>'
  refresh_list_interval: 600s
  resources:
  - resource_query: "resourceType eq 'Microsoft.EventHub/Namespaces'"
    metrics:
    - namespace: Microsoft.EventHub/Namespaces
      name:
        - ActiveConnections
      aggregations:
        - Average
        - Maximum
        - Minimum
      ignore_unsupported: true
      timegrain: PT1M
```

### CODE REVIEW

File: x-pack/metricbeat/module/azure/client_test.go [hunk 1]
```diff
@@ -112,4 +114,157 @@ func TestGetMetricValues(t *testing.T) {
...
+				Timeseries: []*armmonitor.TimeSeriesElement{{
+					Data: []*armmonitor.MetricValue{{
+						Average:   to.Ptr(1.0),
+						Maximum:   to.Ptr(2.0),
+						Minimum:   to.Ptr(3.0),
+						TimeStamp: to.Ptr(time.Now()),
+					}},
+				}},
+				Type:               to.Ptr("Microsoft.Insights/metrics"),
+				Unit:               to.Ptr(armmonitor.MetricUnit("Count")),
```
Comment (gizas) [inline x-pack/metricbeat/module/azure/client_test.go:158]: Wondering if you first save the value of armmonitor.MetricUnit("Count") to a variable and reference here the variable will eliminate this lint error

File: x-pack/metricbeat/module/azure/mock_service.go [hunk 2]
```diff
@@ -43,7 +43,7 @@ func (client *MockService) GetMetricNamespaces(resourceId string) (armmonitor.Me
 
 // GetMetricValues is a mock function for the azure service
 func (client *MockService) GetMetricValues(resourceId string, namespace string, timegrain string, timespan string, metricNames []string, aggregations string, filter string) ([]armmonitor.Metric, string, error) {
-	args := client.Called(resourceId, namespace)
+	args := client.Called(resourceId, namespace, timegrain, timespan, metricNames, aggregations, filter)
```
Comment (gizas) [inline x-pack/metricbeat/module/azure/mock_service.go:46]: You can just add some description what each parameter is here used for
Maybe you can update the  GetMetricValues description

**Comment by gizas** [comment]
LGTM and the tests run successfully on myside

**Review by gizas — APPROVED** [review]

File: x-pack/metricbeat/module/azure/client_test.go (hunk 1 above)
Comment (zmoog) [reply x-pack/metricbeat/module/azure/client_test.go:158]: Oh, I think you're right.

File: x-pack/metricbeat/module/azure/mock_service.go (hunk 2 above)
Comment (zmoog) [reply x-pack/metricbeat/module/azure/mock_service.go:46]: I noticed there is an unused method in the interface. I'll update the docs for the other methods and remove the unused ones in a cleanup PR.

**Pull request merged**